### DISCUSSION TITLE
Diesel for a pocketbase/blitz equivalent

### DISCUSSION THREAD

**Discussion by benwis** [diesel-rs/diesel#4339] [Q&A]
So I've got a small collection of people who want to build an equivalent of pocketbase/blitz for axum+leptos, and I'm trying to evaluate different db libraries to see how close they are to what we need and how much would have to be added on. 

Does Diesel support the following things?
1. Generate/manage a DB schema from a design spec
2. Generate Rust types to easily select/insert on the database itself
3. Drop down to SQL if an operation turns out to perform poorly?

From the docs it looks like 1 is no, but tools exist to represent a database through the table! macro. 2 comes from 1 and 3 looks like yes with [caveats](https://medium.com/@kaustubh.babar/heres-how-to-execute-raw-sql-queries-in-rust-with-diesel-45902931a5f0). Is that fair to say?

**Comment by weiznich** [comment]
So first of all: It's certainly possible to build something like that on diesel. All the necessary information should be there.

To address the single points:

> Generate/manage a DB schema from a design spec

That's kind of possible via the `diesel migration generate --diff-schema` option. The getting started guide and the diesel 2.1 release announcements have more details for that

> Generate Rust types to easily select/insert on the database itself

Diesel does not provide a built-in way to do this, as it decouples rust types from tables. They are designed to map more like 1:1 to queries. That written you can certainly do that (or just use a third party tool like `dsync` for this)

> Drop down to SQL if an operation turns out to perform poorly?

See [`diesel::sql_query`](https://docs.diesel.rs/2.2.x/diesel/fn.sql_query.html). There also several more low level ways to write plain SQL queries, including such that allow you to extend/replace the whole DSL.

**Reply by benwis** [reply]
Thanks weiznich, that's good to kow!
